### PR TITLE
TG-1318 Mgudemann/feature/support arrays in generic parameters

### PR DESCRIPTION
**Pull request by mgudemann**

### CODE REVIEW

File: regression/cbmc-java/generics_array_param1/test.desc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+CORE
+generic.class
+--show-symbol-table
+^EXIT=0$
+^SIGNAL=0$
+^Symbol......: java::GenericType\<java::array\[reference\]of_java::java.lang.Float\>
```
Comment (smowton) [inline regression/cbmc-java/generics_array_param1/test.desc]: AFAIK < and > are not special chars in regexes, so no need to escape

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 2]
```diff
@@ -119,7 +120,25 @@ irep_idt generate_java_generic_typet::build_generic_tag(
     INVARIANT(
       is_java_generic_inst_parameter(param),
       "Only create full concretized generic types");
-    new_tag_buffer << param.subtype().get(ID_identifier);
+    std::string id(id2string(param.subtype().get(ID_identifier)));
```
Comment (smowton) [inline src/java_bytecode/generate_java_generic_type.cpp]: Take by const-ref for the common not-an-array case

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 3]
```diff
@@ -119,7 +120,25 @@ irep_idt generate_java_generic_typet::build_generic_tag(
...
+      // with the same symbols for different instantiations using arrays with
+      // different types
+      if(element_type.id() == ID_pointer)
+      {
+        const symbol_typet element_symbol =
+          to_symbol_type(element_type.subtype());
+        id += "of_" + id2string(element_symbol.get_identifier());
+      }
+    }
+    new_tag_buffer << id;
```
Comment (smowton) [inline src/java_bytecode/generate_java_generic_type.cpp]: Move this above the if-block so you don't need to mutate the string

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 4]
```diff
@@ -119,7 +120,25 @@ irep_idt generate_java_generic_typet::build_generic_tag(
...
+
+      // is it an array of references then we will specialize its identifier
+      // using the type of the objects in the array, else there can be a problem
+      // with the same symbols for different instantiations using arrays with
+      // different types
+      if(element_type.id() == ID_pointer)
+      {
+        const symbol_typet element_symbol =
+          to_symbol_type(element_type.subtype());
+        id += "of_" + id2string(element_symbol.get_identifier());
```
Comment (smowton) [inline src/java_bytecode/generate_java_generic_type.cpp]: Write to `new_tag_buffer` instead of `id`

**Review by smowton — COMMENTED** [review]
Idea is ok, might as well avoid a string copy per usage if we can though

**Comment by mgudemann** [comment]
@smowton thanks, took your comments into account

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 5]
```diff
@@ -119,7 +120,25 @@ irep_idt generate_java_generic_typet::build_generic_tag(
     INVARIANT(
       is_java_generic_inst_parameter(param),
       "Only create full concretized generic types");
-    new_tag_buffer << param.subtype().get(ID_identifier);
+    const std::string &id(id2string(param.subtype().get(ID_identifier)));
+    new_tag_buffer << id;
+    if(has_prefix(id, "java::array["))
+    {
+      namespacet ns(symbol_table);
```
Comment (chrisr-diffblue) [inline src/java_bytecode/generate_java_generic_type.cpp]: I don't think this `namespacet` is used in this block?

**Review by smowton — APPROVED** [review]
Except for @chrisr-diffblue's comment looks good

**Review by chrisr-diffblue — APPROVED** [review]
Apart from the spurious extra `namespacet`, looks good to me too.

File: src/java_bytecode/generate_java_generic_type.cpp (hunk 5 above)
Comment (mgudemann) [reply src/java_bytecode/generate_java_generic_type.cpp]: yes, this was left over from an earlier approach, gone now

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 6]
```diff
@@ -119,7 +119,24 @@ irep_idt generate_java_generic_typet::build_generic_tag(
     INVARIANT(
       is_java_generic_inst_parameter(param),
       "Only create full concretized generic types");
-    new_tag_buffer << param.subtype().get(ID_identifier);
+    const std::string &id(id2string(param.subtype().get(ID_identifier)));
+    new_tag_buffer << id;
+    if(has_prefix(id, "java::array["))
```
Comment (thk123) [inline src/java_bytecode/generate_java_generic_type.cpp]: `java_types.cpp:is_java_array_tag`

File: regression/cbmc-java/generics_array_param1/GenericType.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class GenericType<T> {
```
Comment (thk123) [inline regression/cbmc-java/generics_array_param1/GenericType.java]: As mentioned these should be converted into unit tests added to `generate_concrete_generic_type/generate_java_generic_type.cpp`. As these symbols will no longer be generated except for when an entry function requiring constructing them is.

**Review by thk123 — CHANGES_REQUESTED** [review]

**Comment by mgudemann** [comment]
@chrisr-diffblue @smowton @thk123 due to some changes in `CBMC/develop` this PR changed a bit, please re-review

File: src/java_bytecode/generate_java_generic_type.cpp (hunk 6 above)
Comment (chrisr-diffblue) [reply src/java_bytecode/generate_java_generic_type.cpp]: +1

File: unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp [hunk 8]
```diff
@@ -184,8 +184,9 @@ SCENARIO(
 
   // We want to test that the specialized/instantiated class has it's field
   // type updated, so find the specialized class, not the generic class.
-  const irep_idt test_class=
-    "java::generic_field_array_instantiation$generic<java::array[reference]>";
+  const irep_idt test_class =
+    "java::generic_field_array_instantiation$generic<java::array[reference]of_"
```
Comment (chrisr-diffblue) [inline unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp]: Very minor nit: could you break this string after the `]`, in the same way that you've broken it for `test_class_integer` below, as that seems more readable to me.

Comment (mgudemann) [reply unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp]: ok, but aren't we using `clang-format` to not do this manually anymore?

File: src/java_bytecode/java_utils.cpp [hunk 9]
```diff
@@ -21,11 +21,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
 bool java_is_array_type(const typet &type)
 {
-  if(type.id()!=ID_struct)
+  if(type.id() != ID_struct)
```
Comment (smowton) [inline src/java_bytecode/java_utils.cpp:24]: Don't restyle untouched line

**Review by smowton — APPROVED** [review]

**Comment by smowton** [comment]
Looks good, don't think you should be restyling a line not otherwise touched

File: unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp (hunk 8 above)
Comment (chrisr-diffblue) [reply unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp]: I don't have a strong feeling either way, so feel free to ignore me :-) I just felt the extra break made it more readable. `clang-format` gets around all the 'no space after =` stuff I guess - but its not clear to me how much we're meant to slavishly follow `clang-format`

**Review by chrisr-diffblue — APPROVED** [review]
Looks good, thanks.

File: src/java_bytecode/java_utils.cpp [hunk 10]
```diff
@@ -21,11 +21,9 @@ Author: Daniel Kroening, kroening@kroening.com
...
     return false;
-  return has_prefix(id2string(
-    to_struct_type(type).get_tag()),
-                    "java::array[");
+  return is_java_array_tag(to_struct_type(type).get_tag());
```
Comment (thk123) [inline src/java_bytecode/java_utils.cpp:26]: I didn't know about this method - I wonder if all instances should be replaced with `java_types::is_valid_java_array` (this is out of scope so not required!)

**Review by thk123 — APPROVED** [review]
Thanks for converting them :+1:
